spec/workspace-spec.coffee

```diff
@@ -941,6 +942,100 @@ describe "Workspace", ->
+                callback(searchResult1)
+                new Disposable
+              onDidError: (callback) ->
+                new Disposable
+              onDidSearchPaths: (callback) ->
+                # Invoke the callback with the one notification we plan to send.
+                callback(numPathsToPretendToSearchInCustomDirectorySearcher)
+                new Disposable
+              cancel: ->
+
+            class CustomDirectorySearcher
+              canSearchDirectory: (directory) -> directory.getPath() is dir1
+              search: (directory, options) ->
+                new CustomDirectorySearch
+
+            atom.packages.serviceHub.provide(
+              "atom.directory-searcher", "0.1.0", new CustomDirectorySearcher())
+
+            resultPaths = []
+            onPathsSearched = jasmine.createSpy('onPathsSearched')
+            waitsForPromise ->
+              atom.workspace.scan /aaaa/, {onPathsSearched}, ({filePath}) ->
+                resultPaths.push(filePath)
+
+            runs ->
+              expect(resultPaths.sort()).toEqual([foreignFilePath, file2].sort())
+              # onPathsSearched should be called once by each DirectorySearcher. The order is not
+              # guaranteed, so we can only verify the total number of paths searched is correct
+              # after the second call.
+              expect(onPathsSearched.callCount).toBe(2)
+              expect(onPathsSearched.mostRecentCall.args[0]).toBe(
+                numPathsToPretendToSearchInCustomDirectorySearcher + numPathsSearchedInDir2)
+
+          it "can be cancelled by cancelling one of the DirectorySearchers", ->
+            customDirectorySearcherPromiseInstance = null
+            class CustomDirectorySearchToCancel
+              constructor: ->
+                # Note that hoisting reject in this way is generally frowned upon.
+                @promise = new Promise (resolve, reject) =>
+                  @hoistedReject = reject
+                customDirectorySearcherPromiseInstance = this
+              then: (args...) ->
+                @promise.then.apply(@promise, args)
+              onDidMatch: (callback) ->
+                new Disposable
+              onDidError: (callback) ->
+                new Disposable
+              onDidSearchPaths: (callback) ->
+                new Disposable
+              cancel: ->
+                @hoistedReject()
+
+            class CustomDirectorySearcherToCancel
+              canSearchDirectory: (directory) -> directory.getPath() is dir1
+              search: (directory, options) ->
+                new CustomDirectorySearchToCancel
+
+            atom.packages.serviceHub.provide(
+              "atom.directory-searcher", "0.1.0", new CustomDirectorySearcherToCancel())
+
+            resultPaths = []
+            cancelableSearch = atom.workspace.scan /aaaa/, ({filePath}) ->
+              resultPaths.push(filePath)
+            customDirectorySearcherPromiseInstance.cancel()
```

Comment:
The *individual* `DirectorySearch` objects are not exposed to callers of `Workspace::scan`, right? Currently, this test is verifying the behavior of one of these individual searches. I think that instead, the test should assert is that calling `.cancel()` on the return value of `Workspace::scan` has the effect of calling `cancel` on each individual `DirectorySearch`.